src/mpid/common/shm/mpidu_shm_alloc.c

```diff
@@ -69,8 +69,21 @@ int MPIDU_shm_seg_alloc(size_t len, MPIDU_topo_map_t * map, void **ptr)
 
     MPIR_Assert(segment_len > 0);
 
-    /* allocate an area to check if the segment was allocated symmetrically */
-    segment_len += sizeof(asym_check_region);
+    if (map) {
+        /* set number of processes in segment to 0 */
+        int seg_num_local = 0;
+
+        /* count number of process in the segment */
+        for (int i = 0; i < num_local; i++)
+            if (map->mem_ids[map->seg_id] == map->mem_map[i])
+                seg_num_local++;
+
+        len = (len * seg_num_local) / num_local;        /* update segment length */
+        segment_len = len;
+        segment_len += MPL_MAX(sizeof(asym_check_region), sizeof(MPIDU_shm_seg_md_trl_t));
+    } else
+        /* allocate an area to check if the segment was allocated symmetrically */
+        segment_len += sizeof(asym_check_region);
 
```

Comment:
Re-reading this, probably it would make more sense to push most of this logic (calculation of number of ranks in the segment, and thus segment length) to the caller, leaving the alloc function with the task of allocating the segment and binding it to the requested memory node.

Reply:
How about let the caller do the binding too?

Reply:
I thought about it. But it is actually more clean if the alloc does it. Moreover, since we have to touch the `asym_check_region` at the end of the segment right after the segment is allocated this might cause misplacement of part of the segment into another memory area (DDR by default)

Reply:
Done